### PR TITLE
Updates to InteractionDefinition

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
Overview
---

Adding missing generic "GetType" to def and removing redundant "Tuple parameter

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
Unity also updated the Asset renames from last PR, should be good now

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 1]
```diff
@@ -20,6 +20,14 @@ public InteractionDefinition(uint id, AxisType axisType, DeviceInputType inputTy
             InputType = inputType;
         }
 
+        public InteractionDefinition(uint id, AxisType axisType, Devices.DeviceInputType inputType, InputAction inputAction) : this()
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: We don't need an overload.

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 2]
```diff
@@ -81,11 +89,32 @@ public InteractionDefinition(uint id, AxisType axisType, DeviceInputType inputTy
         /// </summary>
         private Quaternion rotationData;
 
-        private Tuple<Vector3, Quaternion> transformData;
-
         #endregion Definition Data items
 
         #region Get Operators
+		
+		public T GetValue<T>()
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: Formatting

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 3]
```diff
@@ -119,7 +148,7 @@ public Quaternion GetRotation()
 
         public Tuple<Vector3, Quaternion> GetTransform()
         {
-            return transformData;
+            return new Tuple<Vector3, Quaternion>(positionData,rotationData);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: Why change?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 4]
```diff
@@ -81,11 +89,32 @@ public InteractionDefinition(uint id, AxisType axisType, DeviceInputType inputTy
         /// </summary>
         private Quaternion rotationData;
 
-        private Tuple<Vector3, Quaternion> transformData;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: Why remove?

**Comment by StephenHodgson** [comment]
Could we also add the generic setter for now as well so we can test performance?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: I do, to create the interaction for the controller

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: We will most likely always require an action, correct?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 5]
```diff
@@ -126,11 +141,42 @@ public Quaternion GetRotation()
...
+                    boolData = (bool)rawData;
+                    break;
+                case AxisType.SingleAxis:
+                    floatData = (float)rawData;
+                    break;
+                case AxisType.DualAxis:
+                    vector2Data = (Vector2)rawData;
+                    break;
+                case AxisType.ThreeDoFPosition:
+                    positionData = (Vector3)rawData;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: raw data for all set types.

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 6]
```diff
@@ -126,11 +141,42 @@ public Quaternion GetRotation()
...
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
         public void SetValue(object newValue)
         {
             if (AxisType == AxisType.Raw)
             {
-                Changed = newValue != rawData;
+                Changed = newValue.Equals(rawData);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: Changed = !newValue.Equals(rawData);

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 7]
```diff
@@ -184,7 +230,7 @@ public void SetValue(Tuple<Vector3, Quaternion> newValue)
         {
             if (AxisType == AxisType.SixDoF)
             {
-                Changed = newValue.Item1 != transformData.Item1 || newValue.Item2 != transformData.Item2;
+                Changed = transformData == null || newValue.Item1 != transformData.Item1 || newValue.Item2 != transformData.Item2;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]:
```
Changed = newValue == null || transformData == null || newValue.Item1 != transformData.Item1 || newValue.Item2 != transformData.Item2;
```

File: Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        [Test]
+        public void Test02_TestTupleChanged()
+        {
+            var inputDef = new InteractionDefinition(2, AxisType.SixDoF, DeviceInputType.SpatialPointer, new InputAction(1, "Pointer"));
+            var testValue = new Tuple<Vector3, Quaternion>(Vector3.one, new Quaternion(45f, 45f, 45f, 45f));
+
+            inputDef.SetValue(testValue);
+            var setValue = inputDef.GetTransform();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs]: not used?

File: Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        [Test]
+        public void Test04_TestTupleGenericChanged()
+        {
+            var inputDef = new InteractionDefinition(3, AxisType.SixDoF, DeviceInputType.SpatialPointer, new InputAction(1, "Pointer"));
+            var testValue = new Tuple<Vector3, Quaternion>(Vector3.one, new Quaternion(45f, 45f, 45f, 45f));
+
+            inputDef.SetValue<Tuple<Vector3, Quaternion>>(testValue);
+            var setValue = inputDef.GetValue<Tuple<Vector3, Quaternion>>();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs]: not used?

File: Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs (hunk 8 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs]: ok, removed

File: Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs (hunk 9 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit-Tests/InteractionDefinitionTests.cs]: ok, removed

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 10]
```diff
@@ -175,7 +230,7 @@ public void SetValue(Quaternion newValue)
         {
             if (AxisType == AxisType.ThreeDoFRotation)
             {
-                Changed = newValue != rotationData;
+                Changed = !newValue.Equals(rotationData);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: using Equals method incurs boxing for unity types.

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs [hunk 11]
```diff
@@ -184,7 +239,7 @@ public void SetValue(Tuple<Vector3, Quaternion> newValue)
         {
             if (AxisType == AxisType.SixDoF)
             {
-                Changed = newValue.Item1 != transformData.Item1 || newValue.Item2 != transformData.Item2;
+                Changed = (transformData == null && newValue != null) || !transformData.Equals(newValue);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/InteractionDefinition.cs]: Not sure if we're testing all the changed cases properly...

**Pull request merged**